### PR TITLE
Pub sub

### PR DESCRIPTION
**Pull request by VasiliyNovikov**
- Rewritten protocol handling with [System.IO.Pipelines: High performance IO in .NET](https://blogs.msdn.microsoft.com/dotnet/2018/07/09/system-io-pipelines-high-performance-io-in-net/)
- Implemented pub-sub + tests
- Fixed build pipeline flakiness

### CODE REVIEW

File: RedisCore.Tests/RedisClientTestsBase.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        protected static bool HasLocalRedis => !IsVstsBuild || Environment.GetEnvironmentVariable("LOCAL_REDIS") == "true";
+
+        private static readonly int[] BufferSizes = {16, 256, 65536};
+
+        private static IEnumerable<RedisClientConfig>  LocalTestConfigs()
+        {
+            foreach (var bufferSize in BufferSizes)
+            {
+                yield return new RedisClientConfig("127.0.0.1") {BufferSize = bufferSize};
+                //yield return new RedisClientConfig("192.168.0.64");
```
Comment (viktor-ferenczi) [inline RedisCore.Tests/RedisClientTestsBase.cs]: Comment left in code

Comment (VasiliyNovikov) [reply RedisCore.Tests/RedisClientTestsBase.cs]: That's my local redis on Linux VM

Comment (viktor-ferenczi) [reply RedisCore.Tests/RedisClientTestsBase.cs]: Should it go into an environment variable, then defaulting to 127.0.0.1 if it does not present? That could serve as local configuration for your case.

Comment (VasiliyNovikov) [reply RedisCore.Tests/RedisClientTestsBase.cs]: Good idea. Did that

File: RedisCore.Tests/RedisClientTests.cs [hunk 2]
```diff
@@ -113,141 +101,180 @@ public async Task RedisClient_Set_Expire_Test(RedisClientConfig config)
...
+                        for (var j = 0; j < 8; ++j)
+                        {
+                            var testMsg = testMsgBase + j;
+                            await client.Publish(testChannel, testMsg);
+                            var cancellationSource = new CancellationTokenSource(TimeSpan.FromSeconds(0.5));
+                            var actualMsg = await subscription.GetMessage<string>(cancellationSource.Token);
+                            Assert.AreEqual(testMsg, actualMsg);
+                        }
+
+                        await subscription.Unsubscribe();
```
Comment (viktor-ferenczi) [inline RedisCore.Tests/RedisClientTests.cs:208]: This way the user can forget to unsubscribe.
Can it go into Subscription.Dispose somehow?

**Review by viktor-ferenczi — DISMISSED** [review]

File: RedisCore.Tests/RedisClientTests.cs (hunk 2 above)
Comment (VasiliyNovikov) [reply RedisCore.Tests/RedisClientTests.cs:208]: It will in .NET 3.0/ C# 8

**Review by viktor-ferenczi — APPROVED** [review]

**Pull request merged**